src/zjs_event.c

```diff
@@ -40,50 +40,84 @@ void post_event(void* h, jerry_value_t* ret_val)
     }
 }
 
+static uint32_t get_num_events(jerry_value_t emitter)
+{
+    jerry_value_t val = zjs_get_property(emitter, "numEvents");
+    if (!jerry_value_is_number(val)) {
+        ERR_PRINT("emitter had no numEvents property\n");
+        return 0;
+    }
+    uint32_t num = jerry_get_number_value(val);
+    jerry_release_value(val);
+    return num;
+}
+
+static uint32_t get_max_event_listeners(jerry_value_t emitter)
+{
+    jerry_value_t val = zjs_get_property(emitter, "maxListeners");
+    if (!jerry_value_is_number(val)) {
+        ERR_PRINT("emitter had no maxListeners property\n");
+        return 0;
+    }
+    uint32_t num = jerry_get_number_value(val);
+    jerry_release_value(val);
+    return num;
+}
+
+static int32_t get_callback_id(jerry_value_t event_obj)
+{
+    int32_t callback_id = -1;
+    jerry_value_t id_prop = zjs_get_property(event_obj, "callback_id");
+    if (jerry_value_is_number(id_prop)) {
+        // If there already is an event object, get the callback ID
+        zjs_obj_get_int32(event_obj, "callback_id", &callback_id);
+    }
+    jerry_release_value(id_prop);
+
+    return callback_id;
+}
+
 void zjs_add_event_listener(jerry_value_t obj, const char* event,
                             jerry_value_t listener)
 {
-    event_t *ev;
-
     jerry_value_t event_emitter = zjs_get_property(obj, HIDDEN_PROP("event"));
-    if (!jerry_get_object_native_handle(event_emitter, (uintptr_t*)&ev)) {
-        jerry_release_value(event_emitter);
-        ERR_PRINT("native handle not found\n");
+    if (!jerry_value_is_object(event_emitter)) {
+        ERR_PRINT("no event '%s' found\n", event);
         return;
     }
-    jerry_release_value(event_emitter);
+    uint32_t num_events = get_num_events(event_emitter);
+    uint32_t max_listeners = get_max_event_listeners(event_emitter);
 
-    if (ev->num_events >= ev->max_listeners) {
+    if (num_events >= max_listeners) {
+        jerry_release_value(event_emitter);
         ERR_PRINT("max listeners reached\n");
         return;
     }
 
+    jerry_value_t map = zjs_get_property(event_emitter, "map");
+    jerry_value_t event_obj = zjs_get_property(map, event);
+
     // Event object to hold callback ID and eventually listener arguments
-    jerry_value_t event_obj = zjs_get_property(ev->map, event);
     if (!jerry_value_is_object(event_obj)) {
         jerry_release_value(event_obj);
         event_obj = jerry_create_object();
     }
 
-    int32_t callback_id = -1;
-    jerry_value_t id_prop = zjs_get_property(event_obj, "callback_id");
-    if (jerry_value_is_number(id_prop)) {
-        // If there already is an event object, get the callback ID
-        zjs_obj_get_int32(event_obj, "callback_id", &callback_id);
-    }
-    jerry_release_value(id_prop);
-    callback_id = zjs_add_callback_list(listener, obj, ev, post_event,
+    int32_t callback_id = get_callback_id(event_obj);
```

Comment:
if its -1 then add_callback_list will create a new list and return the new callback ID. If its not -1 then it will add the function to the existing callback list.